node/src/components/chain_synchronizer.rs

```diff
@@ -103,50 +169,33 @@ impl ChainSynchronizer {
         self.joining_outcome
     }
 
-    fn start_syncing(
-        &self,
-        effect_builder: EffectBuilder<JoinerEvent>,
-        trusted_hash: BlockHash,
-    ) -> Effects<Event> {
-        info!(%trusted_hash, "synchronizing linear chain");
-        operations::run_chain_sync_task(
-            effect_builder,
-            self.config.clone(),
-            self.metrics.clone(),
-            trusted_hash,
-        )
-        .event(Event::SyncResult)
+    fn fast_sync(&self, effect_builder: EffectBuilder<REv>) -> Effects<Event> {
+        operations::run_fast_sync_task(effect_builder, self.config.clone(), self.metrics.clone())
+            .event(Event::FastSyncResult)
     }
 
-    fn handle_sync_result(
+    fn handle_fast_sync_result(
         &mut self,
-        effect_builder: EffectBuilder<JoinerEvent>,
-        result: Result<BlockHeader, Error>,
+        effect_builder: EffectBuilder<REv>,
+        result: Result<FastSyncOutcome, Error>,
     ) -> Effects<Event> {
         match result {
-            Ok(latest_block_header) => {
-                if latest_block_header.protocol_version() == self.config.protocol_version() {
-                    self.joining_outcome = Some(JoiningOutcome::Synced {
-                        latest_block_header,
-                    });
-                    Effects::new()
-                } else if self
-                    .config
-                    .is_last_block_before_activation(&latest_block_header)
-                {
-                    self.commit_upgrade(effect_builder, latest_block_header)
-                } else {
-                    error!(
-                        ?latest_block_header,
-                        "failed to sync linear chain: unexpected latest block header: not our \
-                         version and not the previous upgrade point"
-                    );
-                    fatal!(
-                        effect_builder,
-                        "unexpected latest block header after syncing"
-                    )
-                    .ignore()
-                }
+            Ok(FastSyncOutcome::ShouldCommitGenesis) => self.commit_genesis(effect_builder),
+            Ok(FastSyncOutcome::ShouldCommitUpgrade {
+                switch_block_header_before_upgrade,
+                is_emergency_upgrade,
+            }) => self.commit_upgrade(
+                effect_builder,
+                switch_block_header_before_upgrade,
+                is_emergency_upgrade,
+            ),
+            Ok(FastSyncOutcome::Synced {
+                highest_block_header,
+            }) => {
+                self.joining_outcome = Some(JoiningOutcome::Synced {
```

Comment:
Yes - I wanted all the logic around determining the joining outcome moved to one place - i.e. into operations.rs, rather than scattered between the two.